### ISSUE TITLE
Question (or Feature Request): Multiple Levels of Variability

### ISSUE THREAD

**Issue by billdenney** [metrumresearchgroup/mrgsolve#505]
I relatively often work with data having multiple levels of variability.  In the current case, I'm working on a model for different organs within an individual where, for example, the left kidney is treated and the right is untreated.

While I could use a structure like inter-occasion variability (https://github.com/mrgsolve/gallery/blob/master/population/iov.md), I was wondering if there is a way to have defined levels of variability so that an inner value is selected for different levels of an outer value.

In `nlme()` terminology, I wonder if something like `e0~1|USUBJID/ORGAN` is possible?  In NONMEM terminology, it would be similar to the use of `$LEVEL`.

**Comment by kylebaron** [comment]
Hi @billdenney - 

There isn't anything baked into mrgsolve to do this.  Just trying to think through the easiest way to get it done. 

Am I understanding the problem right?

We have IDs 1,2,3,4,5,6  and ORGAN = 1 and 2

ID 1,2,3  are all ORGAN = 1
ID 4,5,6 are all ORGAN = 2

so that ID is nested under organ?  Or did I get it reversed?

**Comment by billdenney** [comment]
We have full nesting; the therapy is focused to just the local organ.  So, we have:

IDs 1, 2, 3 and ORGAN 1, 2 such that:

ID | ORGAN 
---- | -------
1 | 1
1 | 2
2 | 1
2 | 2
3 | 1
3 | 2

Though not every ID has both organs, and in some cases, I have up to 6 organs for some IDs.  (It's a surprisingly local therapy; you could think of it like treating each tumor with a specific injection in oncology-- though that's not what it actually is).

**Comment by kylebaron** [comment]
Can you write out what you want to the random effects to be for ID 1 and 2?

**Comment by billdenney** [comment]
Oh, and an important fact for each of the above rows is that I have multiple measurements with each organ (so, you could expand the above table by adding a measure column with ~4 rows per ID/ORGAN pair).

I'm not positive that I understand the question.  Ideally, I'd like to have something like the following for the underlying math:

Row 1: `E = E0 + ETA_ID1_ORGAN1` or `E = E0 + ETA_ID1 + ETA_ID1_ORGAN1` where there is only one value of `ETA_ID1_ORGAN1` per ID=1/ORGAN=1.  And, `ETA_ID1_ORGAN1` is the single value of the random number generator output.  I also do have `ETA_ID1` values that are present in the model, and organs are more correlated within an individual than they are between individuals (e.g. my left and right kidneys are more similar to each other than they are to your left and right kidneys).

In NONMEM, I'm doing the following where "ID" is actually "ORGAN":

```
$INPUT USUBJIDN ID DV <and a lot of other columns>
$THETA
 0.5 ; baseline function
$ETA
 1 ; USUBJID level modifier to baseline function
 2 ; ORGAN level modifier to baseline function
$LEVEL
USUBJID=(1[2])
$PK
BASELINE_EST = THETA(1)+ETA(1)+ETA(2)
```

**Comment by kylebaron** [comment]
Ok.  There isn't anything like `$LEVEL` and I'm trying to understand the mechanics that need to take place to do that. 

Is this what you want?

``` r


library(mrgsolve)
#> 
#> Attaching package: 'mrgsolve'
#> The following object is masked from 'package:stats':
#> 
#>     filter

code <- '

[ param ] 
norgan = 2
organ=1

[ omega ] 
0.04 // Usual ID ETA

[ omega ] 
0.2   // ETA for organ 

[ main ] 
int i = 0;
if(NEWIND==0) {
  double organ1 = ETA(2); 
  simeta();
  double organ2 = ETA(2);
}

capture organ_eta = organ==1 ? organ1 : organ2;
capture id_eta = ETA(1);
capture total_eta = organ_eta + id_eta;

[ capture ] organ
'
data <- expand.grid(amt = 0, cmt = 0, organ = c(1,2), ID = c(1,2),evid=0)
mod <- mcode_cache("level2",code)
#> Building level2 ...
#> done.
mrgsim(mod,data=data, end = -1)
#> Model:  level2 
#> Dim:    4 x 6 
#> Time:   0 to 0 
#> ID:     2 
#>     ID time organ organ_eta  id_eta total_eta
#> 1:   1    0     1    0.1052  0.1817    0.2869
#> 2:   1    0     2   -0.5644  0.1817   -0.3827
#> 3:   2    0     1    0.1052 -0.3847   -0.2795
#> 4:   2    0     2   -0.5644 -0.3847   -0.9490
```

<sup>Created on 2019-09-08 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by billdenney** [comment]
Thank you; that’s exactly what I want!  (And late night on a Saturday, no less!)

In my mind, that is a better solution than the NONMEM method that requires the data to be sorted by ID then organ then time; so non-observation rows must be duplicated when they apply at the ID level (it assumes that levels outside the individual are independent of events and time which is also a good solution for a different problem than mine).  This allows the data to be sorted by ID then time and organ may occur in any order underneath.

For my purposes today, this could be closed.

If you want some additional thoughts on this as a feature request that I accept may make it too complicated to handle within the simulation and could mostly be handled by preprocessing outside of the simulation:

Feature 1: nested random effects outside of an individual:  These effects could be like country-level or site-level effects which are independent of the subject.  (A real example could be typical diet within a region and its effect on diabetes.)  This could have zero or more levels; the levels would be nested but not crossed.

Feature 2: crossed random effects outside an individual.  This is the same as Feature 1, but one individual may have more than one level of the random effect, and the same random effect may be used across individuals.  A real example that I’ve experienced here before is that a device for measurement can have different calibration, and so, each device may give slightly different readings.  Subject 1 may use device 1 on visit 1 and device 2 on visit 2; subject 2 may use device 2 on visit 1 and device 1 on visit 2; subject 3 may use device 3 on visit 1 and device 1 on visit 2; ...

Feature 3: nested random effects within an individual.  This is the organ problem we are discussing.

Feature 4: crossed random effects within an individual.  I don’t think that I’ve experienced a real-world example of this before, but if you implement crosses effects for Feature 2, it seems like there would be no more complexity to implement it within an individual, too, and then it would be here if someone did need it.

As stated above, your answer fully solves my problem today, so please take the features above as thoughts on the problem in an attempt to be helpful and give back a small amount.  If they are out of scope or too complex, the random number generation for all of these situations could be passed in as parameters.

**Comment by billdenney** [comment]
I agree that some R-side prework to put in the random variables at the levels that are like inter-individual variability and higher is the easy workaround solution.  That is what I did for the immediate need.

**Comment by kylebaron** [comment]
I had something similar - ETAs came from a truncated multivariate distribution.  There was a lot of other stuff going on in the model too that required some data customization (really no way around that).  So I added the ETA simulation code to the library of functions that I had developed for the project.  So there's an extra data step that you need to do when you simulate, but I was pretty satisfied with the workflow (was expecting it to feel like more of a burden than it was).
